### PR TITLE
Useful states fixing

### LINKED ISSUE #269: Bug in `Mata::Nfa::reduce`
After the reduction of 
```
@NFA-explicit
%Alphabet-auto
%Initial q0
%Final q4
q0 122 q1
q1 97 q2
q1 97 q4
q1 98 q1
q1 122 q1
q2 97 q1
q2 122 q1
q3 97 q4
```
I get 
```
@NFA-explicit
%Alphabet-auto
%Initial q0
%Final q2
q0 122 q1
q1 97 q2
q1 98 q1
q1 122 q1
```
which is not equivalent.

### PR DESCRIPTION
**Pull request by vhavlena**
This PR fixes the `useful_states` function with the Tarjan-based implementation. A couple of questions to consider/discuss:

- Tarjan-based SCC decomposition is quite general and can be used elsewhere. I was wondering if we can do some parameterized Tarjan by a lambda performing action on each SCC. In this framework, however, it is not easily possible to express the computation of useful states as the usefulness is not a property of a single SCC.
- The working with iterators inside `NodeData` sucks. This could be somehow improved by `Post::transitions` iterator (iterator over a single post getting triplets (q,a,p))
- There is a lot of obsolete code related to trim and computation of useful states. I remove it later in this PR.

### CODE REVIEW

File: src/nfa/nfa.cc [hunk 1]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+        }
+        if(rec_call) continue;
+
+        // check if we have the root of a SCC
+        if(act_state_data.lowlink == act_state_data.index) {
+            State st;
+            bool final_scc = false;
+            std::vector<State> scc;
+            do {
+                st = tarjan_stack.back();
```
Comment (tfiedor) [inline src/nfa/nfa.cc:455]: I wonder, can `tarjan_stack` be empty at this line? Have you tried it on some extreme cases? Empty Nfa or maybe One State Nfa?

File: src/nfa/nfa.cc [hunk 2]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+    State get_act_succ() {
+        align_succ();
+        return *(this->targets_it);
+    }
+
+    void move_next_succ() {
+        if(this->post_it == this->post_end) {
+            return;
+        }
+        this->targets_it++;
```
Comment (tfiedor) [inline src/nfa/nfa.cc]:
```suggestion
        ++this->targets_it;
```
Might be a teensy-weensy-bit faster.

**Review by tfiedor — APPROVED** [review]
Seems fine, though, looking at tests and code, I'm wondering if it might fail on some extreme or corner cases of NFA.

File: include/mata/nfa/nfa.hh [hunk 3]
```diff
@@ -205,6 +205,8 @@ public:
 
     BoolVector get_useful_states() const;
 
+    BoolVector get_useful_states_tarjan() const;
```
Comment (jurajsic) [inline include/mata/nfa/nfa.hh]: Why did you introduce new function instead of rewriting `get_useful_states()`?

Comment (jurajsic) [reply include/mata/nfa/nfa.hh]: Also some comment what this function returns would be nice (I assume BoolVector for states representing which are useful).

File: src/nfa/nfa.cc [hunk 4]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
     };
 };
 
+struct NodeData {
```
Comment (jurajsic) [inline src/nfa/nfa.cc]: What does this struct represent? Some comment would be nice. Also the name should probably make it more explicit that it is connected with computing useful states.

File: include/mata/nfa/nfa.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/nfa/nfa.hh]: We should keep both the original function and the Tarjan one in my opinion. But the original function will only call the Tarjan implementation. If we want to keep the classical approach, we should rename it accordingly (something like `get_useful_states_classic()`, `get_useful_states_naive()`, etc.) which would be called from the `get_useful_states_dispatcher` if we decided to not use the Tarjan algorithm, for some reason.

If we think we will only ever use the Tarjan approach, then we should rename the original function as Juraj suggests.

File: tests/nfa/nfa.cc [hunk 5]
```diff
@@ -3043,3 +3043,90 @@ TEST_CASE("Mata::Nfa:: print_to_mata") {
 
 	CHECK(are_equivalent(aut_big, aut_big_from_mata));
 }
+
+TEST_CASE("Mata::Nfa::trim bug") {
+	Nfa aut(5, {0}, {4});
+    aut.delta.add(0, 122, 1);
```
Comment (Adda0) [inline tests/nfa/nfa.cc]:
```suggestion
	Nfa aut(5, {0}, {4});
       aut.delta.add(0, 122, 1);
```
Fix indentation

File: src/nfa/nfa.cc (hunk 4 above)
Comment (Adda0) [inline src/nfa/nfa.cc]: This should be kept in an anonymous namespace because users are not supposed to use this structure manually.

File: src/nfa/nfa.cc [hunk 6]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+    Post::const_iterator post_it;
+    Post::const_iterator post_end;
+    StateSet::const_iterator targets_it{};
+    StateSet::const_iterator targets_end{};
+    unsigned long index;
+    unsigned long lowlink;
+    bool initilized;
+    bool on_stack;
```
Comment (Adda0) [inline src/nfa/nfa.cc]: Could we add a comment explaining what is the purpose of these member variables?

File: src/nfa/nfa.cc [hunk 7]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+        : post_it(delta[q].cbegin()), post_end(delta[q].cend()), index(index), lowlink(index), initilized(true), on_stack(true) {
+        if (post_it != post_end) {
+            targets_it = post_it->cbegin();
+            targets_end = post_it->cend();
+        }
+    };
+
+    // TODO: this sucks. In fact, if you want to check that you have the last sucessor, you need to 
+    // first align the iterators.
+    // TODO: this is super-ugly. If we introduce Post::transitions iterator, this could be much easier.
```
Comment (Adda0) [inline src/nfa/nfa.cc]: It seems that we agreed to create the iterator over transitions. How will it improve the code here if we have to allocate each new transition `Trans` every iteration?

File: src/nfa/nfa.cc [hunk 8]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+                ++this->post_it;
+                if (this->post_it != this->post_end) {
+                    this->targets_it = this->post_it->cbegin();
+                    this->targets_end = this->post_it->cend();
+                }
+            } 
+        }
+    }
+
+    State get_act_succ() {
```
Comment (Adda0) [inline src/nfa/nfa.cc]: `act` as in `active`? Would `get_current_successor()`, or in the shortened form, `get_curr_succ()` better express what this function gets you?

File: src/nfa/nfa.cc [hunk 9]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+        return this->post_it == this->post_end;
+    }
+};
+
+BoolVector Nfa::get_useful_states_tarjan() const {
+    BoolVector reached_and_reaching(this->size(),false); 
+    std::vector<NodeData> node_info(this->size());
+    std::deque<State> program_stack;
+    std::deque<State> tarjan_stack;
+    unsigned long cnt = 0;
```
Comment (Adda0) [inline src/nfa/nfa.cc]: What are we counting? Either rename the variable, or add a comment, please.

**Review by Adda0 — CHANGES_REQUESTED** [review]
The PR looks good overall. I opened a few conversations I would like to resolve before merging. 

Do we have some performance testing results for this optimization? How did it compare?

File: include/mata/nfa/nfa.hh (hunk 3 above)
Comment (vhavlena) [reply include/mata/nfa/nfa.hh]: I remove the original function (and a couple of other obsolete code related to trim) later in this PR. The original code is left basically for Lukas for a comparison (there was some discussion whether it could be simplified somehow according to the original function).

File: src/nfa/nfa.cc (hunk 7 above)
Comment (vhavlena) [reply src/nfa/nfa.cc]: The code will be simpler.

File: src/nfa/nfa.cc (hunk 8 above)
Comment (vhavlena) [reply src/nfa/nfa.cc]: Whatever

**Comment by vhavlena** [comment]
> The PR looks good overall. I opened a few converasitons I would like to resolve before merging.
> 
> Do we have some performance testing results for this optimization? How did it compare?

No, we don't. But it is not necessary at this point. It is a bug fixing, not an optimization, so I am not sure, how do you imagine the performance testing.

File: src/nfa/nfa.cc (hunk 7 above)
Comment (Adda0) [reply src/nfa/nfa.cc]: That is true. But it might be potentially slower, right? I do not understand how you and Lukáš intend to iterate over triples without allocating them.

**Comment by Adda0** [comment]
You are right, my bad. We should just try to run perf tests for a few examples then to be sure that there is no extreme performance degradation with the rewrite of the function in case that it is slower. I cannot reliably guess on how it will perform.

Just comparing the original and the fixed version performance should be enough here, I think.

File: include/mata/nfa/nfa.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/nfa/nfa.hh]: Then what about creating a dispatcher function and hiding the implementation details in a separate source file where we can experiment with different implementations?

> I remove the original function ... later in this PR. The original code is left basically for Lukas for a comparison ...

I do not understand. The original code is still there, but are do we think we will ever use it again? If no, we can remove the code entirely. And ideally, we should think about the potential simplification as a part of this PR, before merging it. Or do you mean that the original code is not removed yet and will be removed before merging the PR? Then it would be OK with me.

File: src/nfa/nfa.cc [hunk 10]
```diff
@@ -312,6 +312,136 @@ struct StackLevel {
...
+
+    bool is_end() {
+        align_succ();
+        return this->post_it == this->post_end;
+    }
+};
+
+BoolVector Nfa::get_useful_states_tarjan() const {
+    BoolVector reached_and_reaching(this->size(),false); 
+    std::vector<NodeData> node_info(this->size());
```
Comment (jurajsic) [inline src/nfa/nfa.cc]: `NodeData` contains `bool` and `unsigned long` variables which are not initialized, I think this will lead to UB.

Comment (vhavlena) [reply src/nfa/nfa.cc]: I don't get it

**Comment by vhavlena** [comment]
> Just comparing the original and the fixed version performance should be enough here, I think.

Sorry, I won't do it. Comparing performance of the unsound procedure (there was a fundamental problem, not a minor fix) with the fixed version does not make sense to me.

File: src/nfa/nfa.cc (hunk 1 above)
Comment (vhavlena) [reply src/nfa/nfa.cc:455]: Should not, in theory. That's all I can say. I can put more tests there.

**Comment by Adda0** [comment]
> >   Just comparing the original and the fixed version performance should be enough here, I think.
>
> Sorry, I won't do it. Comparing performance of the unsound procedure (there was a fundamental problem, not a minor fix) with the fixed version does not make sense to me.

Fair enough.

File: src/nfa/nfa.cc (hunk 10 above)
Comment (jurajsic) [reply src/nfa/nfa.cc]: You are creating vector of `NodeData`, and they use default constructor. Will that initialize the `bool` members to `false`? You assume it does, on line 383 you expect `false`.

**Comment by vhavlena** [comment]
> > > Just comparing the original and the fixed version performance should be enough here, I think.
> > 
> > 
> > Sorry, I won't do it. Comparing performance of the unsound procedure (there was a fundamental problem, not a minor fix) with the fixed version does not make sense to me.
> 
> Fair enough.

It is used in `trim` operation. So, after merge we can try it in noodler to check some degradation/bugs.

File: src/nfa/nfa.cc (hunk 1 above)
Comment (tfiedor) [reply src/nfa/nfa.cc:455]: Can you please write a theoretical proof, that this cannot happen? That should suffice. (that is just joke). You are much smarter than me, so I will trust you.

File: src/nfa/nfa.cc (hunk 10 above)
Comment (vhavlena) [reply src/nfa/nfa.cc]: I expected that it is initialized to `false` by default. Is it the case? @Adda0

File: src/nfa/nfa.cc (hunk 1 above)
Comment (vhavlena) [reply src/nfa/nfa.cc:455]: I guess you will regret 😄

Comment (tfiedor) [reply src/nfa/nfa.cc:455]: Or give it to Adam, he can find failing corner cases while taking a nap.

File: include/mata/nfa/nfa.hh (hunk 3 above)
Comment (vhavlena) [reply include/mata/nfa/nfa.hh]: Fine, I remove it now. Lukas can find the original version in the git history.

File: src/nfa/nfa.cc (hunk 7 above)
Comment (vhavlena) [reply src/nfa/nfa.cc]: Yes. But if the difference will be negligible, I suggest to use the transition iteration in order to keep the code at least a bit readable.

File: src/nfa/nfa.cc (hunk 10 above)
Comment (Adda0) [reply src/nfa/nfa.cc]: The magic around default initialization, value initialization, explicit initialization, and more is complicated. For a brief overview, see example in [LearnCpp article on default initialization](https://www.learncpp.com/cpp-tutorial/default-member-initialization/). The best approach is to initialize everything as explicitly as possible, at least with `{}` in member variable declaration. In this case, both I and Clang-tidy believe that there is a possibility of UB with the default constructor in the following members: `index`, `lowlink`, `initilized`, `on_stack`.

Another great explanation is in the following [blog post](https://arne-mertz.de/2015/08/new-c-features-default-initializers-for-member-variables/). Initialize everything everywhere and you will be safe. If your member variable is not a class, it will not get initialized if you use only `NodeData node_data;`.

File: src/nfa/nfa.cc (hunk 7 above)
Comment (Adda0) [reply src/nfa/nfa.cc]: Definitely. It would be so much more readable, indeed. I consider the conversation resolved. When we implement the transition iterator, we can think about this again.

File: src/nfa/nfa.cc (hunk 10 above)
Comment (vhavlena) [reply src/nfa/nfa.cc]: Great, thanks!

File: src/nfa/nfa.cc [hunk 11]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
+            if (post_it != post_end) {
+                targets_it = post_it->cbegin();
+                targets_end = post_it->cend();
+            }
+        };
+
+        // TODO: this sucks. In fact, if you want to check that you have the last sucessor, you need to 
+        // first align the iterators.
+        // TODO: this is super-ugly. If we introduce Post::transitions iterator, this could be much easier.
+        void align_succ() {
```
Comment (kilohsakul) [inline src/nfa/nfa.cc:330]: Add some comment on what it does?

File: src/nfa/nfa.cc [hunk 12]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
-#else
-    std::vector<StackLevel> stack;//the DFS stack
-    //tracking elements seems to cost more than it saves, switching it off
-    BoolVector reached(size(),false); // Reachable from initial state.
-    BoolVector reached_and_reaching(size(),false); // Reachable from initial state and reaches final state.
-#endif
+BoolVector Nfa::get_useful_states() const {
+    BoolVector reached_and_reaching(this->size(),false); 
+    std::vector<TarjanNodeData> node_info(this->size());
+    std::deque<State> program_stack;
```
Comment (kilohsakul) [inline src/nfa/nfa.cc:400]: Explain program stack plz.
Is it the program stack of that code at wiki? 
The link to that code and some explanation of how it is related might be good somewhere here?

File: src/nfa/nfa.cc [hunk 13]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
-    StackLevel(State q, const Delta & delta) : state(q), post_it(delta[q].cbegin()), post_end(delta[q].cend()) {
-        if (post_it != post_end) {
-            targets_it = post_it->cbegin();
-            targets_end = post_it->cend();
+        State get_curr_succ() {
+            align_succ();
+            return *(this->targets_it);
+        }
+
+        void move_next_succ() {
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: Should it be move_to_nex_succ  or something like that?

File: src/nfa/nfa.cc [hunk 14]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
+            if(this->post_it == this->post_end) {
+                return;
+            }
+            ++this->targets_it;
+        }
+
+        bool is_end() {
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: no_more_succ ? or something like that?

File: src/nfa/nfa.cc [hunk 15]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
+        } else { // return from the recursive call
+            State act_succ = act_state_data.get_curr_succ();
+            act_state_data.lowlink = std::min(act_state_data.lowlink, node_info[act_succ].lowlink);
+            // act_succ is the state that cased the recursive call. Move to another successor.
+            act_state_data.move_next_succ();
+        }
+
+        // iterate through outgoing edges
+        State next_state;
+        bool rec_call = false;
```
Comment (kilohsakul) [inline src/nfa/nfa.cc:430]: can you explain rec_call?

File: src/nfa/nfa.cc [hunk 16]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
+                node_info[st].on_stack = false;
+
+                // SCC contains a final state
+                if(reached_and_reaching[st]) {
+                    final_scc = true;
                 }
+                scc.push_back(st);
+            } while(st != act_state);
+            if(final_scc) {
+                for(const State& st : scc) reached_and_reaching[st] = true;
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: I thought that one does not have to collect the scc and iterate, that just going down the stack on hitting a final state and marking everything final would be fine. Maybe I would need an explanation.

File: include/mata/nfa/nfa.hh (hunk 3 above)
Comment (kilohsakul) [reply include/mata/nfa/nfa.hh]: Just keep this and remove everything else I think.

File: src/nfa/nfa.cc (hunk 7 above)
Comment (kilohsakul) [reply src/nfa/nfa.cc]: Lets not allocate a transition every time. The iterator can have an internal transition, just one, ++ will update its components, and dereference will return a const reference to it. Then one just reads the components he wants from there, or coppies it if neccessary.

**Comment by kilohsakul** [comment]
We need a carefull friendly explanation of the algorithm, it is tricky and not described enywhere. Please give it some energy.
Perhaps some explanation with respect to the code at wiki, if it helps ... ?

Otherwise it is wondreful, of course :).

File: src/nfa/nfa.cc [hunk 17]
```diff
@@ -237,11 +237,11 @@ void Nfa::trim_reverting(StateToStateMap* state_map)
...
 #ifdef _STATIC_STRUCTURES_
-    BoolVector useful_states{ get_useful_states() };
+    BoolVector useful_states{ useful_states() };
     useful_states.clear();
-    useful_states = get_useful_states();
+    useful_states = useful_states();
 #else
-    BoolVector useful_states{ get_useful_states() };
+    BoolVector useful_states(get_useful_states());
 #endif
```
Comment (kilohsakul) [inline src/nfa/nfa.cc:245]: I would just kill the static structures everywhere, including here, I don't think playing with that until tacas is worth it.

File: src/nfa/nfa.cc [hunk 18]
```diff
@@ -295,84 +295,142 @@ Nfa Nfa::get_trimmed_automaton(StateToStateMap* state_map) const {
...
+            State st;
+            bool final_scc = false;
+            std::vector<State> scc;
+            do {
+                st = tarjan_stack.back();
+                tarjan_stack.pop_back();
+                node_info[st].on_stack = false;
+
+                // SCC contains a final state
+                if(reached_and_reaching[st]) {
```
Comment (kilohsakul) [inline src/nfa/nfa.cc]: btw is reached_and_reaching a good name? I remember someone complaining about it earlier. It could also be "usefull".

**Comment by vhavlena** [comment]
> We need a carefull friendly explanation of the algorithm, it is tricky and not described enywhere. Please give it some energy. Perhaps some explanation with respect to the code at wiki, if it helps ... ?
> 
> Otherwise it is wondreful, of course :).

Ok, I write down something

**Review by Adda0 — APPROVED** [review]
Well done. I am happy with the PR, and it seems to work as intended. The PR is ready for merge from where I stand. Let me know when you are finished with the PR.

File: src/nfa/nfa.cc [hunk 19]
```diff
@@ -237,11 +237,11 @@ void Nfa::trim_reverting(StateToStateMap* state_map)
 void Nfa::trim_inplace(StateToStateMap* state_map)
 {
 #ifdef _STATIC_STRUCTURES_
-    BoolVector useful_states{ get_useful_states() };
+    BoolVector useful_states{ useful_states() };
     useful_states.clear();
-    useful_states = get_useful_states();
+    useful_states = useful_states();
```
Comment (jurajsic) [inline src/nfa/nfa.cc:242]: Is there supposed to be `useful_states()` or `get_useful_states()`?

**Review by jurajsic — APPROVED** [review]
Looks good.

File: src/nfa/nfa.cc (hunk 19 above)
Comment (vhavlena) [reply src/nfa/nfa.cc:242]: 🤷

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:1]: It is ok, the new explanation is good, thanks, just please anwer somehow the comments on names of variables and static structures plz.

**Comment by vhavlena** [comment]
Has anyone some other issue? Or can we merge it?

**Comment by tfiedor** [comment]
> Has anyone some other issue? Or can we merge it?

I have some issues with groin joints. Nothing for your PR, though.

File: src/nfa/nfa.cc (hunk 12 above)
Comment (vhavlena) [reply src/nfa/nfa.cc:400]: It should be commented in the method preface

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:1]: Please check it

**Comment by Adda0** [comment]
We should merge this PR to fix the bug in `devel`. @kilohsakul Do you approve the additional changes made to the PR?

**Review by kilohsakul — APPROVED** [review]

**Comment by kilohsakul** [comment]
> We should merge this PR to fix the bug in `devel`. @kilohsakul Do you approve the additional changes made to the PR?

ok yesyes
